rbac/management/group/definer.py

```diff
@@ -144,14 +157,62 @@ def add_roles(group, roles_or_role_ids, tenant, user=None):
                     "into groups."
                 )
                 raise serializers.ValidationError({key: _(message)})
+
         # Only add the role if it was not attached
-        if not system_policy.roles.filter(pk=role.pk).exists():
-            system_policy.roles.add(role)
+        if system_policy.roles.filter(pk=role.pk).exists():
+            continue
 
-            # Send notifications
-            group_role_change_notification_handler(user, group, role, "added")
+        system_policy.roles.add(role)
+        tuples_to_add: list[Relationship] = []
 
+        if role.system:
+            try:
+                mapping = (
+                    BindingMapping.objects.select_for_update()
+                    .filter(
+                        role=role,
+                        resource_type_namespace="rbac",
+                        resource_type_name="workspace",
+                        # TODO: don't use org id once we have workspace built ins
+                        resource_id=tenant.org_id,
+                    )
+                    .get()
+                )
+                tuples_to_add.append(mapping.add_group_to_bindings(group.uuid))
+                mapping.save(force_update=True)
+            except BindingMapping.DoesNotExist:
+                id = str(uuid4())
+                binding = V2rolebinding(
+                    id,
+                    V2role(role.uuid, True, frozenset()),
+                    # TODO: don't use org id once we have workspace built ins
+                    V2boundresource(("rbac", "workspace"), tenant.org_id),
+                    groups=[group.uuid],
+                )
+                mapping = BindingMapping.for_role_binding(binding, role)
+                mapping.save(force_insert=True)
+                tuples_to_add.extend(mapping.as_tuples())
+        else:
+            # Intentionally queried this way to avoid use of cached binding_mappings
+            # There is a risk of phantom reads here, but this is why we lock the custom Role
+            bindings: Iterable[BindingMapping] = BindingMapping.objects.filter(role=role).select_for_update()
+            for mapping in bindings:
+                tuples_to_add.append(mapping.add_group_to_bindings(group.uuid))
+                mapping.save(force_update=True)
 
+        # Send notifications
+        # TODO: maybe use top-level method instead so replicator can be modified for tests
+        OutboxReplicator(role).replicate(
```

Comment:
Is intent to have one replication event for each role change ?
pros:
- payload will not grow 
cons:
- probably harder failure handling when one event will fail in source connector ?